In [10]:
import pandas as pd
import numpy as np
import os

In [12]:
path = r'/Users/christina/Desktop/Instcart Basket Analysis'

In [14]:
# Import dataset orders_products_prior.csv
df_ords_prior = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [15]:
# Import dataset orders_cleaned.csv
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_cleaned.csv'), index_col = False)

In [16]:
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge.pkl'))

1.Create "price_label" column.

In [18]:
# creat a sebset of the dataframe
df = ords_prods_merge[:1000000]

In [19]:
def price_label (row):
    if row['prices'] <= 5:
        return 'low-range product'
    elif row['prices'] > 5 and (row ['prices'] <= 15):
        return 'mid-range product'
    elif row ['prices'] > 15:
        return 'High range'
    else : return ' not enough data'

In [20]:
df ['price_range'] = df.apply(price_label,axis = 1)

/var/folders/9s/ltmjqdbd1hv01gxpsxjdzt1w0000gn/T/ipykernel_91388/2951760581.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df ['price_range'] = df.apply(price_label,axis = 1)


In [21]:
df.shape

(1000000, 20)

In [22]:
df['prices'].max()

24.5

In [23]:
df['price_range'].value_counts(dropna = False)

price_range
mid-range product    652638
low-range product    338018
High range             9344
Name: count, dtype: int64

In [24]:
df['prices'].value_counts(dropna = False)

prices
11.3    53649
9.0     44266
3.8     32376
1.4     31093
4.9     30112
        ...  
24.5       17
16.9       11
19.9       11
18.6        6
18.8        6
Name: count, Length: 157, dtype: int64

In [25]:
df['price_range'].max()

'mid-range product'

In [26]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/9s/ltmjqdbd1hv01gxpsxjdzt1w0000gn/T/ipykernel_91388/1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [27]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [28]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [29]:
ords_prods_merge. loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [30]:
ords_prods_merge. loc [(ords_prods_merge ['prices']) <= 15 & (ords_prods_merge ['prices'] > 5), 'price_range_loc'] = 'Mid_range product'

In [31]:
ords_prods_merge.loc [ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [32]:
ords_prods_merge['price_range_loc'].value_counts (dropna = False)

price_range_loc
NaN                   21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [34]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [35]:
result = []

In [36]:
for value in ords_prods_merge ['orders_day_of_week'] :
    if value == 0:
        result.append ('busiest_day')
    elif value == 4:
        result.append ('least_busy')
    else:
        result.append('regular_day')

In [38]:
result

['regular_day',
 'regular_day',
 'busiest_day',
 'regular_day',
 'least_busy',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'least_busy',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'busiest_day',
 'busiest_day',
 'busiest_day',
 'busiest_day',
 'busiest_day',
 'least_busy',
 'regular_day',
 'regular_day',
 'busiest_day',
 'regular_day',
 'regular_day',
 'least_busy',
 'regular_day',
 'busiest_day',
 'regular_day',
 'busiest_day',
 'least_busy',
 'busiest_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'regular_day',
 'busiest_day',
 'regular_day',
 'regular_day',
 'busiest_day',
 'least_busy',
 'regular_day',
 'regular_day',
 'regular_day'

1. create 'busiest_day' columns

In [46]:
# Change "order_dow" column name to "order_day_of_week."
ords_prods_merge.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [48]:
# Find frequency of orders_day_of_week.
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [60]:
# Use for-loop to determine whether orders are on "busiest day" (0 = Saturday), "least busy" (4 = Wednesday), or "regularly busy" (other days of the week).

result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [61]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [66]:
# Create new column in ords_prods_merge to view results in context.
ords_prods_merge['busiest_day'] = result

In [68]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

2. Update "busiest day" column to "busiest days" to identify order day of the week as "Busiest days," "Slowest days," or "Regularly busy."

In [71]:
# Create column “busiest_days" from "orders_day_of_week" to identify two busiest days (0, 1), 
# slowest days (4, 3), or regularly busy (all other days).

result_2 = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [73]:
result_2

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',

3. Check values of new "busiest days" column for accuracy with observations in markdown format.

In [76]:
# Create new 'busiest_days' column in ords_prods_merge to view results in context.

ords_prods_merge['busiest_days'] = result_2

In [78]:
ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [80]:
# Check output of ords_prods_merge with new "busiest days" column.

ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0_y  order_id  user_id eval_set  ...  \
0             19     5.8          1987   3139998      138    prior  ...   
1             19     5.8          1989   1977647      138    prior  ...   
2             19     5.8         11433    389851      709    prior  ...   
3             19     5.8         12198    652770      764    prior  ...   
4             19     5.8         12200   1813452      764    prior  ...   

   order_hour_of_day  days_since_prior_order  first_order  add_to_cart_order  \
0                 11                     3.0        False                  5   
1                 17                    20.0        False                  1   
2                 21                     6.0        False                 20   
3                 13                     NaN         True                 10   
4                 17                     9.0        False                 11   

   reordered  _merge  _merge2 price_range_loc     busiest_day    busiest_days  
0          0    both     both             NaN  Regularly busy  Regularly busy  
1          1    both     both             NaN  Regularly busy  Regularly busy  
2          0    both     both             NaN     Busiest day    Busiest days  
3          0    both     both             NaN  Regularly busy    Slowest days  
4          1    both     both             NaN      Least busy    Slowest days  

[5 rows x 22 columns]

The total value counts for the ords_prods_merge dataframe is equal 
to the total sum of all groupings (i.e. "Regularly busy" + "Busiest 
days" + "Slowest days"). In addition, the listed sums for each label
match the sum totals for the qualifying days of the week (e.g. Total
"Busiest days" = 11864412, which matches orders_day_of_week 0 + 1, or 6204182 + 5660230)

4. Create new column "busiest_period_of_day" to identify time periods “Most orders,” “Average orders,” and “Fewest orders.”

In [85]:
# Check value counts in "order_hour_of_day" column.

ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

The value counts listed above will be split into equal thirds for the following 
labels: "Most orders" = 10, 11, 14, 15, 13, 12, 16, and 9; "Fewest orders" = 23, 6, 0, 1, 5, 2, 4, 
and 3; "Average orders" = all remaining values in order_hour_of_day.

In [89]:
# Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_3 = []

for value in ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [91]:
result_3

['Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most ord

In [93]:
ords_prods_merge['busiest_period_of_day'] = result_3

5. Print frequency of "busiest_period_of_day" column.

In [96]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [98]:
ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

OSError: [Errno 28] No space left on device